crates/builder/src/graph/adj_list.rs

```diff
@@ -3,22 +3,29 @@ use crate::{
     CsrLayout, DirectedDegrees, DirectedNeighbors, DirectedNeighborsWithValues, Graph, Target,
     UndirectedDegrees, UndirectedNeighbors, UndirectedNeighborsWithValues,
 };
+use crate::{EdgeMutation, EdgeMutationWithValues};
 
 use log::info;
-use std::sync::Mutex;
+use std::sync::{Mutex, MutexGuard};
 use std::time::Instant;
 
 use crate::graph::csr::NodeValues;
 use rayon::prelude::*;
 
 #[derive(Debug)]
 pub struct AdjacencyList<NI, EV> {
-    edges: Vec<Vec<Target<NI, EV>>>,
+    edges: Vec<Mutex<Vec<Target<NI, EV>>>>,
```

Comment:
With the addition of the Mutex it would be interesting to statically check that certain traits are or are not implemented, i.e. Send+Sync